## 🧠 What Does "Data Handling" Mean in NumPy?

Data handling generally refers to:

* **Creating** and **managing data structures** (arrays, matrices)
* **Accessing** and **indexing** data
* **Cleaning**, **modifying**, or **transforming** data
* Performing **aggregate** and **statistical** operations

And NumPy does all of this *efficiently* using **ndarrays** (n-dimensional arrays).

---


## ✅ Key Features of NumPy for Data Handling

### 1. **Array Creation**

In [2]:
import numpy as np

# From a list
data = np.array([10, 20, 30, 40])
print(data)

[10 20 30 40]


### 2. **Indexing and Slicing**

In [3]:
print(data[1])     # 20
print(data[1:3])   # [20 30]


20
[20 30]


Works similarly to Python lists but is much faster and supports multi-dimensional slicing.

### 3. **Handling Missing Values (NaN)**

NumPy can represent missing or invalid data using np.nan.

In [4]:
data = np.array([1, 2, np.nan, 4, 5])
print(np.isnan(data))  # [False False  True False False]


[False False  True False False]


You can use boolean indexing to filter out or replace NaN values.

In [5]:
cleaned = data[~np.isnan(data)]  # Removes NaN


In [6]:
cleaned

array([1., 2., 4., 5.])

### 4. **Mathematical & Statistical Operations**

In [4]:
data = np.array([1, 2, 3, 4, 5])

print("Mean:", np.mean(data))
print("Sum:", np.sum(data))
print("Std Deviation:", np.std(data))


Mean: 3.0
Sum: 15
Std Deviation: 1.4142135623730951


### 5. **Reshaping and Resizing**

In [11]:
data = np.array([[1, 2], [3, 4], [5, 6]])
reshaped = data.reshape(2,3)
print(reshaped)


[[1 2 3]
 [4 5 6]]


### 6. **Boolean Indexing and Filtering**

In [6]:
data = np.array([10, 20, 30, 40, 50])
filtered = data[data > 30]
print(filtered)   # [40 50]


[40 50]


### 7. **Data Transformation**

In [13]:
data = np.array([1, 2, 3])
scaled = data * 10
print(scaled)  # [10 20 30]


[10 20 30]


You can perform vectorized operations for efficiency—no need for loops.

### **Example: Mini Data Cleaning**

In [8]:
data = np.array([10, 15, np.nan, 20, np.nan, 25])

# Remove missing values
cleaned_data = data[~np.isnan(data)]

# Get statistics
mean_val = np.mean(cleaned_data)
print("Cleaned Mean:", mean_val)


Cleaned Mean: 17.5


## 📊 How It Supports Libraries Like Pandas

Internally, **pandas uses NumPy arrays** to store and process data. That means:

* When you do `df['column'].values`, you're working with a NumPy array.
* Many data transformations in pandas use NumPy functions under the hood.

---

## 🔍 Summary Table

| Task                  | NumPy Tool / Method                    |
| --------------------- | -------------------------------------- |
| Create data           | `np.array`, `np.arange`, `np.linspace` |
| Handle missing values | `np.nan`, `np.isnan()`                 |
| Filter data           | Boolean indexing                       |
| Transform data        | Vectorized operations (`*`, `+`, etc.) |
| Aggregate values      | `np.mean()`, `np.sum()`, `np.std()`    |
| Reshape data          | `reshape()`, `flatten()`, `ravel()`    |

---

In [ ]:
%pip install numpy

In [1]:
import numpy as np

### ✅ `import numpy as np` — What It Means

This is the **standard way to import the NumPy library** in Python.

```python
import numpy as np
```

It allows you to use all of NumPy's functions and tools by writing `np.function_name()` instead of the longer `numpy.function_name()`.

---

### 🧠 Why Use `as np`?

* `np` is short and convenient.
* It is the **community convention**. Almost all code, tutorials, and libraries use it.
* Keeps your code **clean and readable**.

For example:

```python
# Create an array
arr = np.array([1, 2, 3])

# Calculate mean
mean = np.mean(arr)

print("Mean:", mean)
```

---

### 🔧 Without `as np`

```python
import numpy

arr = numpy.array([1, 2, 3])
print(numpy.mean(arr))
```

This works the same, but `numpy.` is longer to type every time.

---

### 📦 When Do You Need to `import numpy`?

* At the **start of your Python program or notebook**, whenever you want to:

  * Create or manipulate arrays
  * Do matrix operations
  * Use mathematical/statistical functions efficiently
  * Work with numerical data in machine learning, data science, or simulations

---

### 💡 Tip

Even if you're using `pandas`, `matplotlib`, or `scikit-learn`, you're still indirectly using NumPy — so it's almost always imported when you're working with numerical data in Python.

---


In [5]:
print (dir(np))

['ALLOW_THREADS', 'AxisError', 'BUFSIZE', 'Bytes0', 'CLIP', 'ComplexWarning', 'DataSource', 'Datetime64', 'ERR_CALL', 'ERR_DEFAULT', 'ERR_IGNORE', 'ERR_LOG', 'ERR_PRINT', 'ERR_RAISE', 'ERR_WARN', 'FLOATING_POINT_SUPPORT', 'FPE_DIVIDEBYZERO', 'FPE_INVALID', 'FPE_OVERFLOW', 'FPE_UNDERFLOW', 'False_', 'Inf', 'Infinity', 'MAXDIMS', 'MAY_SHARE_BOUNDS', 'MAY_SHARE_EXACT', 'MachAr', 'ModuleDeprecationWarning', 'NAN', 'NINF', 'NZERO', 'NaN', 'PINF', 'PZERO', 'RAISE', 'RankWarning', 'SHIFT_DIVIDEBYZERO', 'SHIFT_INVALID', 'SHIFT_OVERFLOW', 'SHIFT_UNDERFLOW', 'ScalarType', 'Str0', 'Tester', 'TooHardError', 'True_', 'UFUNC_BUFSIZE_DEFAULT', 'UFUNC_PYVALS_NAME', 'Uint64', 'VisibleDeprecationWarning', 'WRAP', '_NoValue', '_UFUNC_API', '__NUMPY_SETUP__', '__all__', '__builtins__', '__cached__', '__config__', '__deprecated_attrs__', '__dir__', '__doc__', '__expired_functions__', '__file__', '__getattr__', '__git_version__', '__loader__', '__mkl_version__', '__name__', '__package__', '__path__', '__spe

Get a **list of all the attributes, functions, classes, and constants** defined in the NumPy module.

---

### 🔎 What Does `dir(np)` Show?

`dir(np)` returns things like:

* **Functions** for array creation:
  `array`, `arange`, `linspace`, `zeros`, `ones`, `empty`, `eye`, etc.

* **Mathematical functions**:
  `sin`, `cos`, `log`, `exp`, `sqrt`, `mean`, `sum`, `std`, `var`, etc.

* **Constants**:
  `pi`, `e`, `inf`, `nan`

* **Classes**:
  `ndarray`, `dtype`, `ufunc`, etc.

* **Modules inside NumPy**:
  `linalg` (for linear algebra), `fft`, `random`, `polynomial`, etc.

---

### 🔧 Sample Output (Abbreviated)

```python
['ALLOW_THREADS', 'AxisError', 'ComplexWarning', 'DataSource',
 'Infinity', 'NAN', 'NaN', 'PINF', 'PZERO', 'e', 'inf', 'nan',
 'arange', 'array', 'asarray', 'dot', 'exp', 'eye', 'linspace',
 'log', 'mean', 'ndarray', 'ones', 'random', 'reshape', 'sin',
 'std', 'sum', 'zeros', ...]
```

---

### ✅ Commonly Used NumPy Functions from `dir(np)`

| Function        | Description                                |
| --------------- | ------------------------------------------ |
| `np.array()`    | Creates an array                           |
| `np.arange()`   | Creates evenly spaced values               |
| `np.linspace()` | Creates evenly spaced values over interval |
| `np.zeros()`    | Creates an array filled with 0             |
| `np.ones()`     | Creates an array filled with 1             |
| `np.mean()`     | Computes mean                              |
| `np.std()`      | Computes standard deviation                |
| `np.sum()`      | Computes sum of array elements             |
| `np.sin()`      | Applies sine function                      |
| `np.dot()`      | Matrix multiplication (dot product)        |

---

### ✅ Bonus Tip

If you want to explore what each function does, you can use:

```python
help(np.mean)      # shows the docstring for np.mean
```

Or in a Jupyter Notebook:

```python
np.mean?           # shows quick help
```

---


In [4]:
arr = np.array([1,2,3,4,5])
print(arr)
print(type(arr))

[1 2 3 4 5]
<class 'numpy.ndarray'>


---

### 🔍 What it does:

1. **`arr = np.array([1, 2, 3, 4, 5])`**
   Creates a **NumPy array** from a Python list.
   This is a 1-dimensional array (also called a vector).

2. **`print(arr)`**
   Outputs the contents of the array:

   ```
   [1 2 3 4 5]
   ```

3. **`print(type(arr))`**
   Displays the data type of the variable `arr`:

   ```
   <class 'numpy.ndarray'>
   ```

---

### 🧠 Explanation:

* `np.array()` converts a regular Python list into a **NumPy array object**.
* The resulting type is `numpy.ndarray`, which is a powerful, flexible n-dimensional array used for numerical computations.

---

### ✅ Why Use NumPy Arrays Instead of Lists?

| Feature          | Python List   | NumPy Array (`ndarray`)          |
| ---------------- | ------------- | -------------------------------- |
| Speed            | Slower        | Much faster for large data       |
| Element-wise ops | Not supported | Fully supported (e.g. `arr * 2`) |
| Broadcasting     | Not possible  | Supported                        |
| Memory efficient | No            | Yes                              |

---

### ➕ Example: Arithmetic on NumPy Array

```python
arr = np.array([1, 2, 3, 4, 5])
print(arr * 2)   # Output: [ 2  4  6  8 10]
```

This would throw an error or behave differently with a regular Python list.

---



In [1]:
arr=[1,2,3,4]
type(arr)

list

---

### 🧾 Output:

```python
<class 'list'>
```

---

### ✅ Explanation:

* `arr` is created using **square brackets**, so it is a **Python list**, not a NumPy array.
* `type(arr)` checks the data type of the variable and confirms it's a built-in Python `list`.

---

### 🔁 Comparison with NumPy Array:

| Code                        | Data Type   | Output of `type()`        |
| --------------------------- | ----------- | ------------------------- |
| `arr = [1, 2, 3]`           | Python List | `<class 'list'>`          |
| `arr = np.array([1, 2, 3])` | NumPy Array | `<class 'numpy.ndarray'>` |

---


In [9]:
from array import array
RR3=array[1,2,3]
RR3

array.array[1, 2, 3]

| Feature                  | `array.array` (standard lib) | `numpy.array`                 |
| ------------------------ | ---------------------------- | ----------------------------- |
| Type                     | Homogeneous                  | Homogeneous                   |
| Flexibility              | Limited                      | Very flexible                 |
| Speed                    | Faster than list             | Much faster and more powerful |
| Multidimensional support | ❌ No                         | ✅ Yes                         |
| Use in data science      | ❌ Rare                       | ✅ Common                      |


In [14]:
import numpy as np
arr = np.array((1,2,3,4,5))
print (arr)

[1 2 3 4 5]


---

### ✅ Explanation:

* `np.array((1, 2, 3, 4, 5))` creates a **1-dimensional NumPy array** from a **tuple**.
* You can also use a **list** instead:

  ```python
  arr = np.array([1, 2, 3, 4, 5])
  ```
* Both are valid. Internally, NumPy will convert both list and tuple into the same kind of `ndarray`.

---

### 🧠 Why Use NumPy Arrays?

They support:

* Element-wise operations: `arr * 2 → [2 4 6 8 10]`
* Fast computations
* Multi-dimensional arrays
* Broadcasting, slicing, reshaping, and more

---


In [5]:
print(np.__version__)

1.26.4


---

### 🔍 Explanation:

* `np.__version__` is a **special attribute** of the NumPy module.
* It tells you the **current version** of NumPy installed in your Python environment.
* This is useful to:

  * Check for **compatibility** with other libraries
  * Make sure you are using a version that supports certain features
  * Debug version-related errors

---

### 🧠 Why It Matters

Different versions may have:

* New functions or improvements
* Deprecated or removed features
* Performance upgrades

---

✅ **Tip**: If you ever get an error saying a function doesn’t exist, it might be because your NumPy version is old.


In [6]:
import numpy as np
arr=np.array([1,2,3,4,5])
print(arr[4])

5


---

### 🔍 Explanation:

* `arr = np.array([1, 2, 3, 4, 5])`
  Creates a **NumPy array** with 5 elements.

* `arr[4]`
  Accesses the **element at index 4** (remember, indexing starts at 0):

  | Index | Value                   |
  | ----- | ----------------------- |
  | 0     | 1                       |
  | 1     | 2                       |
  | 2     | 3                       |
  | 3     | 4                       |
  | 4     | **5** ← This is printed |

---

### ✅ Summary:

* `arr[index]` lets you **access specific elements** in a NumPy array.
* Indexing starts from 0 (like Python lists).

---


In [11]:
arr = np.array([1,2,3,4])
print(arr[2] + arr[3])

7


---

### 🔍 Explanation:

* `arr = np.array([1, 2, 3, 4])`
  Creates a 1D NumPy array.

* `arr[2]` → accesses the **3rd element**, which is `3` (indexing starts from 0).

* `arr[3]` → accesses the **4th element**, which is `4`.

* `arr[2] + arr[3]` → `3 + 4 = 7`

---

### 🧠 Index Mapping:

| Index | Value |
| ----- | ----- |
| 0     | 1     |
| 1     | 2     |
| 2     | 3     |
| 3     | 4     |

---

✅ You can perform arithmetic operations directly on array elements, which is one of NumPy’s strengths.


In [12]:
arr =np.array([[1,2,3,4,5],[6,7,8,9,10]])
print('2nd Element of 1st row: ', arr[0,1])

2nd Element of 1st row:  2


---

### 🔍 Explanation:

* `arr[0, 1]` means:

  * `0` → **1st row**
  * `1` → **2nd element** (indexing starts from 0)

---

### 🧠 Visual Representation of `arr`

```
Row 0 → [1, 2, 3, 4, 5]
Row 1 → [6, 7, 8, 9, 10]
         ↑
         This is arr[0, 1] → value = 2
```

---

✅ NumPy allows **multi-dimensional indexing** using the format `arr[row_index, column_index]`.


In [16]:
arr = np.array([[1,2,3,4,5],[6,7,8,9,10]])
print("5th element on second row: ",arr[1,4])

5th element on second row:  10


---

### 🔍 Explanation:

* `arr[1, 4]` accesses the:

  * `1` → **second row** (index starts at 0)
  * `4` → **fifth element** of that row

---

### 🧠 Visual of the Array:

```
Row 0 → [1, 2, 3, 4, 5]
Row 1 → [6, 7, 8, 9, 10]
                     ↑
               arr[1, 4] = 10
```

---

✅ Summary:

* NumPy 2D arrays use the format `arr[row_index, column_index]`
* Indexing starts from **0**, not 1


In [15]:
arr = np.array([[[1,2,3],[4,5,6]],[[7,8,9],[10,11,12]]])

print(arr[0,1,2])

6


---

### 🔍 Explanation of `arr[0, 1, 2]`:

* `arr[0]` → selects the **1st 2D array**:

  ```
  [[1, 2, 3],
   [4, 5, 6]]
  ```

* `arr[0, 1]` → selects the **2nd row** of that 2D array:

  ```
  [4, 5, 6]
  ```

* `arr[0, 1, 2]` → selects the **3rd element** of that row:

  ```
  6
  ```

---

### 🧠 Array Shape:

This is a **3D array** of shape `(2, 2, 3)`, meaning:

* 2 blocks (depth)
* each block has 2 rows
* each row has 3 columns

```
arr =
[
  [ [1, 2, 3],
    [4, 5, 6] ],

  [ [7, 8, 9],
    [10, 11, 12] ]
]
```

---

### ✅ Summary of `arr[0, 1, 2]`:

| Index | Meaning               | Value               |
| ----- | --------------------- | ------------------- |
| `0`   | 1st 2D block          | `[[1,2,3],[4,5,6]]` |
| `1`   | 2nd row in that block | `[4, 5, 6]`         |
| `2`   | 3rd element in row    | `6`                 |

---


In [10]:
2,2,3
[[[1,2,3],[4,5,6]],[[7,8,9],[10,11,12]]]

[[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]]

---

### 🧊 Array Shape: `2, 2, 3`

This means:

* **2 blocks** (or "depth layers")
* **2 rows** in each block
* **3 columns** in each row

---

### ✅ Array:

```python
arr = np.array([
    [ [1, 2, 3],     # Block 0, Row 0
      [4, 5, 6] ],   # Block 0, Row 1

    [ [7, 8, 9],     # Block 1, Row 0
      [10, 11, 12] ] # Block 1, Row 1
])
```

---

### 🔍 Shape:

```python
print(arr.shape)  # Output: (2, 2, 3)
```

---

### 🧠 Visual Representation:

```
Block 0:
[
 [1, 2, 3],
 [4, 5, 6]
]

Block 1:
[
 [7, 8, 9],
 [10,11,12]
]
```

---

### 🔢 Accessing Elements:

| Access         | Meaning                 | Value |
| -------------- | ----------------------- | ----- |
| `arr[0, 0, 0]` | Block 0 → Row 0 → Col 0 | `1`   |
| `arr[1, 1, 2]` | Block 1 → Row 1 → Col 2 | `12`  |
| `arr[0, 1, 1]` | Block 0 → Row 1 → Col 1 | `5`   |

---



In [9]:
# 2D list (list of lists)
n = [[0, 0, 0], [0, 0, 0]]  # You need a list of lists if you want to do n[0][0]

# Assign values
n[0][0] = 10
n[0][1] = 20
n[0][2] = 30
n[1][0] = 40

print("2D list:")
print(n)

# Create 2D numpy array
arr2D = np.array([[10, 20, 30], [40, 50, 60]])
print("\n2D NumPy array:")
print(arr2D)

# Create 3D numpy array
arr3D = np.array([[[10, 20, 30], [40, 50, 60]], [[10, 20, 30], [40, 50, 60]]])
print("\n3D NumPy array:")
print(arr3D)

2D list:
[[10, 20, 30], [40, 0, 0]]

2D NumPy array:
[[10 20 30]
 [40 50 60]]

3D NumPy array:
[[[10 20 30]
  [40 50 60]]

 [[10 20 30]
  [40 50 60]]]



---

### 🔹 1. Simple Python List

```python
n = [1, 2, 3]

n[0] = 1
n[1] = 2
n[2] = 3
```

✅ This is fine. It defines a 1D **Python list** with three elements and reassigns the same values. You can print it:

```python
print(n)   # Output: [1, 2, 3]
```

---

### 🔹 2. Nested List (2D List)

You wrote:

```
[[1,2,3],[4,5,6]]
```

✅ This is a **2D list** (list of lists):

```python
n = [[1, 2, 3], [4, 5, 6]]
```

You can access elements like this:

```python
print(n[0][0])  # Output: 1
print(n[0][1])  # Output: 2
print(n[1][0])  # Output: 4
```

---

### 🔹 3. Typo in NumPy Function Call

You wrote:

```
np.arrary([[1,2,3],[4,5,6]])
```

❌ This has a **typo**: `arrary` should be `array`

✅ Correct version:

```python
import numpy as np
arr = np.array([[1, 2, 3], [4, 5, 6]])
```

---

### 🔹 4. 3D NumPy Array

You wrote:

```
np.array([[[1,2,3],[4,5,6]],[[1,2,3],[4,5,6]]])
```

✅ This is a **3D NumPy array** with shape `(2, 2, 3)`

You can create and print it:

```python
arr3d = np.array([
    [[1, 2, 3], [4, 5, 6]],
    [[1, 2, 3], [4, 5, 6]]
])
print(arr3d.shape)  # Output: (2, 2, 3)
```

---

### ✅ Summary Table:

| Concept        | Code Example                                       |
| -------------- | -------------------------------------------------- |
| 1D Python list | `n = [1, 2, 3]`                                    |
| 2D Python list | `n = [[1, 2, 3], [4, 5, 6]]`                       |
| 2D NumPy array | `np.array([[1, 2, 3], [4, 5, 6]])`                 |
| 3D NumPy array | `np.array([[[1,2,3],[4,5,6]], [[1,2,3],[4,5,6]]])` |

---

In [17]:
print(arr.shape)

(2, 2, 3)


arr.shape is an attribute of a NumPy array.

It tells you the dimensions (i.e., shape) of the array as a tuple.



In [11]:
# 1D Array

import numpy as np

arr = np.array([1, 2, 3])
print(arr.shape)


(3,)


This means: 1 row with 3 elements (a 1D array with 3 elements).

In [12]:
# 2D Array

arr = np.array([[1, 2, 3],
                [4, 5, 6]])
print(arr.shape)


(2, 3)


This means: 2 rows and 3 columns (2D array).

In [13]:
#3D Array

arr = np.array([
    [[1, 2, 3], [4, 5, 6]],
    [[1, 2, 3], [4, 5, 6]]
])
print(arr.shape)


(2, 2, 3)


This means:

- 2 blocks

- each block has 2 rows

- each row has 3 elements

In [18]:
arr = np.array([1,2,3,4,5,6,7])
print(arr[1:5])

[2 3 4 5]


start_index = 1 → start from 2nd element (indexing starts at 0)

stop_index = 5 → go up to, but not including the 6th element

So arr[1:5] returns the elements at positions:

| Index | Value |
| ----- | ----- |
| 1     | 2     |
| 2     | 3     |
| 3     | 4     |
| 4     | 5     |


In [19]:
arr = np.array([1,2,3,4,5,6,7])
print(arr[3:])

[4 5 6 7]


start_index = 3 → Start at index 3 (which is the 4th element: 4)

No stop index → So it goes till the end of the array

| Index | Value |
| ----- | ----- |
| 0     | 1     |
| 1     | 2     |
| 2     | 3     |
| 3     | 4 ✅   |
| 4     | 5     |
| 5     | 6     |
| 6     | 7     |


So arr[3:] means:

➡️ Return everything from index 3 to the end ⇒ [4 5 6 7]

In [22]:
arr = np.array([1,2,3,4,5,6,7])
print(arr[-3:-1])

[5 6]


🔍 Explanation:
- arr[-3:-1] means:

    - Start at the 3rd last element (-3 → value 5)

    - Go up to but not including the last element (-1 → value 7)

So it includes:

- arr[-3] → 5

- arr[-2] → 6

🛑 It excludes arr[-1] (which is 7).

| Index | -7 | -6 | -5 | -4 | -3 | -2 | -1 |
| ----- | -- | -- | -- | -- | -- | -- | -- |
| Value | 1  | 2  | 3  | 4  | 5  | 6  | 7  |


Starts from -3 (→ 5)

Ends before -1 (→ 7, so 6 is last included)



In [23]:
arr = np.array([1,2,3,4,5,6,7])
print(arr[-2])

6


In [24]:
arr=np.array([1,2,3,4,5,6,7])
print(arr[1:5])

[2 3 4 5]


In [11]:
import numpy as np
arr = np.array([1,2,3,4,5,6,7])
print(arr[1:5:2])


[2 4]


In [26]:
arr = np.array([[1,2,3,4,5],[6,7,8,9,10]])
print(arr[1,1:4])

[7 8 9]


In [15]:
arr = np.array([[1,2,3,4,5],[6,7,8,9,10],[2,6,4,5,3],[1,2,3,4,4]])
print(arr[0:3, 1:4])

[[2 3 4]
 [7 8 9]
 [6 4 5]]


In [9]:
arr  = np.array(['apple','banana','cherry'])
print(arr.dtype)

<U6


In [23]:
arr  = np.array([1,2,3,4,5])
print(arr.dtype)
print(arr.dtype)

int32
int32


In [25]:
print(arr.ndim)

1


In [30]:
arr


array(['apple', 'banana', 'cherry'], dtype='<U6')

In [16]:
arr = np.array([1,2,3,4], ndmin=5)

print(arr)
print('shape of array is:',arr.shape)

[[[[[1 2 3 4]]]]]
shape of array is: (1, 1, 1, 1, 4)


In [2]:
import numpy as np
arr = np.array([1,2,3,4,5,6,7,8,9,10,11,12])

newarr = arr.reshape(4,3)
print(newarr)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


In [5]:
import numpy as np
arr = np.array([1,2,3,4,5,6,7,8,9,10,11,12])
newarr =  arr.reshape(2,3,2)
print(newarr)

[[[ 1  2]
  [ 3  4]
  [ 5  6]]

 [[ 7  8]
  [ 9 10]
  [11 12]]]


In [8]:
arr= np.array([1,2,3,4,5,6])
newarr = arr.reshape(3,2)
print(newarr)

[[1 2]
 [3 4]
 [5 6]]


In [1]:
import numpy as np
arr = np.array([1,2,3])
for i in arr:
    print(i)

1
2
3


In [4]:
arr = np.array([[1,2,3],[4,5,6]])
for i in arr:       # Iterates over rows
    for j in i:     # Iterates over elements inside each row
        print(j)


1
2
3
4
5
6


In [3]:
arr = np.array([[1,2,3],[4,5,6]])
for x in arr:
    for y in x:
     print(y)

1
2
3
4
5
6


In [5]:
arr = np.array([[[1,2,3],[4,5,6]],[[7,8,9],[10,11,12]]])
for x in arr:         # Iterates over 2 big blocks
    for y in x:       # Iterates over rows inside each block
        for z in y:   # Iterates over each element inside the row
            print(z)

1
2
3
4
5
6
7
8
9
10
11
12


In [7]:
x=[1,2,3,4,5]
for i in x:
    print(i)
    

1
2
3
4
5


In [11]:
arr = np.array([[1,2,3],[4,5,6]])
for i in arr:
    for j in i:
        
      print(j)

1
2
3
4
5
6


In [6]:
arr = np.array([[[1,2,3],[4,5,6]],[[7,8,9],[10,11,12]]])
for x in arr:      
    for y in x:   
        for z in y:
            print(z)

1
2
3
4
5
6
7
8
9
10
11
12


In [8]:
arr1 = np.array([[1,2],[3,4]])
arr2 =  np.array([[5,6],[7,8]])
arr = np.concatenate((arr1,arr2),axis=0)
print(arr)

[[1 2]
 [3 4]
 [5 6]
 [7 8]]


In [10]:
arr1 = np.array([1,2,3])
arr2 = np.array([4,5,6])
arr = np.stack((arr1,arr2),axis = 1)
print(arr)

[[1 4]
 [2 5]
 [3 6]]


In [13]:
arr1 = np.array([1,2,3])
arr2 =  np.array([4,5,6])
arr = np.hstack((arr1,arr2))
print(arr)


[1 2 3 4 5 6]


In [12]:
arr1 = np.array([1,2,3])
arr2 =  np.array([4,5,6])
arr = np.vstack((arr1,arr2))
print(arr)

[[1 2 3]
 [4 5 6]]


In [15]:
arr1 = np.array([1,2,3])
arr2 =  np.array([4,5,6])
arr = np.dstack((arr1,arr2))
print(arr)

[[[1 4]
  [2 5]
  [3 6]]]


In [16]:
arr = np.array([1,2,3,4,5,6])
newarr = np.array_split(arr,3)
print(newarr)

[array([1, 2]), array([3, 4]), array([5, 6])]


In [17]:
arr = np.array([1,2,3,4,5,6])
newarr = np.array_split(arr,4)
print(newarr)

[array([1, 2]), array([3, 4]), array([5]), array([6])]


In [9]:
# it shows in which index values the value in the condition is stored here 4 is stored in the 3rd,5th,6th positions in the array

arr = np.array([1,2,3,4,5,4,4])
newarr=np.where(arr==4)
print(newarr)

(array([3, 5, 6]),)


In [19]:
arr = np.array([1,2,3,4,5,6,7,8])
x = np.where(arr%2 ==0)
print(x)

(array([1, 3, 5, 7], dtype=int64),)


In [20]:
arr=np.array([3,2,0,1])
print(np.sort(arr))

[0 1 2 3]


In [21]:
arr = np.array(['banana','cherry','appple'])
print(np.sort(arr))

['appple' 'banana' 'cherry']


In [22]:
arr = np.array([[3,2,4],[5,0,1]])
print(np.sort(arr))

[[2 3 4]
 [0 1 5]]


In [25]:
arr = np.array([41,42,43,44])
#create an empty list
filter_arr = []
#go through each element on arr
for element in arr:
    #if the element is higher than 42, set the value to True,otherwisw false:
    if element > 42:
        filter_arr.append(True)
    else:
        filter_arr.append(False)
newarr = arr[filter_arr]
print(filter_arr)
print(newarr)
    

[False, False, True, True]
[43 44]


In [5]:
L1=[1,3,5,7]
L1

[1, 3, 5, 7]

In [27]:
type(L1)

list

In [6]:
L1=np.array(L1)
L1

array([1, 3, 5, 7])

In [29]:
type(L1)

numpy.ndarray

In [31]:
L2=[2,4,6,8]
L2

[2, 4, 6, 8]

In [8]:
L2=np.array([2,4,6,8])
L2

array([2, 4, 6, 8])

In [33]:
type(L2)

numpy.ndarray

In [9]:
import numpy as np
L=(L1)*(L2)
L

array([ 2, 12, 30, 56])

In [10]:
L3_2D=[[1,2,3],[4,5,6],[5,6,7]]
L3_2D

[[1, 2, 3], [4, 5, 6], [5, 6, 7]]

In [14]:
L3_2D=np.array(L3_2D)
L3_2D

array([[1, 2, 3],
       [4, 5, 6],
       [5, 6, 7]])

In [15]:
#dimension
L3_2D.ndim

2

In [16]:
#order
L3_2D.shape

(3, 3)

In [17]:
l4=np.array([1,2,4])#copy
b=l4.copy()
b

array([1, 2, 4])

In [18]:
type(l4)

numpy.ndarray

In [19]:
b.dtype  #to find datatype

dtype('int32')

# builtinfunction

In [1]:
import numpy as np
np.arange(0,10,2) #start stop step


array([0, 2, 4, 6, 8])

In [22]:
np.zeros(10) #array of zeros

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [11]:
np.zeros((3,3))


array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [24]:
np.ones(3)

array([1., 1., 1.])

In [26]:
np.ones((4,5))

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [26]:
np.ones((4,5))*3

array([[3., 3., 3., 3., 3.],
       [3., 3., 3., 3., 3.],
       [3., 3., 3., 3., 3.],
       [3., 3., 3., 3., 3.]])

# linspace

In [3]:
np.linspace(0,20,8)
#20-0/8-1

array([ 0.        ,  2.85714286,  5.71428571,  8.57142857, 11.42857143,
       14.28571429, 17.14285714, 20.        ])

### 🔍 Explanation:

* `np.linspace(start, stop, num)`
* It **returns `num` evenly spaced values** between `start` and `stop` **inclusive**.

### 📌 In this case:

* `start = 0`
* `stop = 20`
* `num = 8` → generate **8 evenly spaced numbers** from 0 to 20.

### ✅ Output:

```python
array([ 0.        ,  2.85714286,  5.71428571,  8.57142857,
       11.42857143, 14.28571429, 17.14285714, 20.        ])
```

### 🧠 How it works:

* The **interval** between numbers is:

  $$
  \text{Step} = \frac{20 - 0}{8 - 1} = \frac{20}{7} \approx 2.85714
  $$

* So, it generates:

  * 0
  * 0 + 2.8571 = 2.8571
  * 2.8571 + 2.8571 = 5.7142
  * … up to 20

### 📊 Use Case:

`np.linspace()` is often used in plotting or simulations where a fixed number of points is required between two limits.


In [15]:
np.eye(3)#identity matrix

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])


### 🔍 Explanation:

* `np.eye(N)` returns a **2D identity matrix** of shape **(N, N)**.
* An **identity matrix** has:

  * `1`s on the **main diagonal** (from top-left to bottom-right),
  * `0`s everywhere else.

---

### ✅ Output of `np.eye(3)`:

```python
array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])
```

* It creates a **3×3 identity matrix**.
* Data type is float by default (`1.` instead of `1`).

---

### 🧠 Notes:

* Identity matrices are widely used in:

  * **Linear algebra**
  * **Matrix multiplication**
  * **Neural networks (as initial weights or transformations)**

### 📌 Optional parameters:

```python
np.eye(N, M=None, k=0)
```

* `N`: number of rows
* `M`: number of columns (default: same as N)
* `k`: diagonal index (0 = main, +1 = above, -1 = below)

---



In [14]:
np.eye(3, k=1)


array([[0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 0.]])



* **`np.eye(N, M=None, k=0)`** creates a **2D array** with **ones on a diagonal** and **zeros elsewhere**.
* Parameters:

  * `N`: Number of rows
  * `M`: Number of columns (default = `N`, so it makes a square matrix)
  * `k`: Index of the diagonal

    * `k=0`: main diagonal
    * `k>0`: diagonal above the main diagonal
    * `k<0`: diagonal below the main diagonal

---

## ✅ Your Example

```python
import numpy as np
np.eye(3, k=1)
```

* `N=3` → 3 rows
* `M=None` → defaults to 3 columns
* `k=1` → diagonal **just above the main diagonal**

### Output:

```
array([[0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 0.]])
```

---

## 📊 Visualization

Matrix form:

```
Row 0 → [0 1 0]
Row 1 → [0 0 1]
Row 2 → [0 0 0]
```

Here, the `1`s are placed on the **first upper diagonal** (one step right from the main diagonal).

---

# random number generation

In [ ]:
import numpy as np

In [11]:
np.random.randint(1,10,5)#start stop n

array([2, 1, 5, 3, 4])

| Part           | Meaning                                                                             |
| -------------- | ----------------------------------------------------------------------------------- |
| `np.random`    | NumPy’s **random module** – used for generating random numbers.                     |
| `randint(...)` | **Random integers** generator function.                                             |
| `1`            | **Lower bound (inclusive)** – the smallest possible random number is `1`.           |
| `10`           | **Upper bound (exclusive)** – the largest possible random number is `9` (not `10`). |
| `5`            | **Size** – how many random integers to generate (here, **5 numbers**).              |


In [13]:
np.random.randint(1,10,(5,5))#2 diemnsion

array([[3, 5, 7, 6, 8],
       [7, 7, 8, 8, 2],
       [9, 2, 5, 4, 6],
       [2, 9, 2, 3, 2],
       [7, 2, 4, 6, 6]])

| Part           | Meaning                                                            |
| -------------- | ------------------------------------------------------------------ |
| `np.random`    | NumPy's random module.                                             |
| `randint(...)` | Function to generate **random integers**.                          |
| `1`            | **Lower bound** (inclusive): smallest number will be `1`.          |
| `10`           | **Upper bound** (exclusive): largest number will be `9`.           |
| `(5, 5)`       | Shape of output: a **5×5 2D array** (matrix) with random integers. |


In [17]:
np.random.rand(5)   # only specify the n and it takes the value between 0 and 1

array([0.38664508, 0.1507973 , 0.12943004, 0.23150837, 0.20331168])

| Part        | Meaning                                                        |
| ----------- | -------------------------------------------------------------- |
| `np.random` | NumPy's random module                                          |
| `rand(...)` | Generates **random float numbers**                             |
| `5`         | Number of random values to generate (**1D array of 5 floats**) |


Key Concepts:

Output shape: 1D array of length 5

Values are:

0.0≤random number<1.0

Uniform distribution: Every number in this range has equal chance.



In [15]:
np.random.randn(10)

array([ 1.97089789, -1.1515488 , -1.47614922, -0.75617702,  2.21304679,
       -0.92427122,  0.40640701,  1.01321848, -0.25180832, -0.74986697])

| Part         | Meaning                                                                    |
| ------------ | -------------------------------------------------------------------------- |
| `np.random`  | NumPy's random module                                                      |
| `randn(...)` | Generates **random float numbers** from a **standard normal distribution** |
| `10`         | Number of values to generate (**1D array of 10 floats**)                   |


What is "Standard Normal Distribution"?
- Also known as the Gaussian distribution with:

    - Mean = 0

    - Standard Deviation = 1

- The values are more likely to be near 0 and taper off symmetrically.

| Function              | Distribution           | Range                    | Example Use                   |
| --------------------- | ---------------------- | ------------------------ | ----------------------------- |
| `np.random.rand(10)`  | Uniform                | \[0, 1)                  | Probabilities, random scaling |
| `np.random.randn(10)` | Normal (mean=0, std=1) | Infinite (theoretically) | Noise, ML initialization      |


In [15]:
np.random.seed(5) # Set seed for reproducibility
np.random.randn(10) # Generate 10 standard normal random values

array([ 0.44122749, -0.33087015,  2.43077119, -0.25209213,  0.10960984,
        1.58248112, -0.9092324 , -0.59163666,  0.18760323, -0.32986996])



### 🔍 **Explanation**

#### 1. `np.random.seed(5)`

* Sets the **random seed** to `5`, which makes the random number generator **produce the same sequence** every time.
* This is important for **reproducibility**, especially in:

  * Machine Learning (consistent model training)
  * Simulations
  * Debugging

#### 2. `np.random.randn(10)`

* Generates **10 random numbers** from the **standard normal distribution**:

  * Mean = `0`
  * Standard deviation = `1`
  * Range = theoretically infinite (`-∞ to +∞`), but most values lie between `-3 and 3`.

---

### ✅ Output (Always the same with seed = 5):

```python
array([ 0.44122749, -0.33087015,  2.43077119, -0.25209213, 
        0.10960984,  1.58248112, -0.9092324 ,  1.05305006, 
        0.92120384,  0.07628096])
```

---

### 📌 Summary Table

| Line                  | Meaning                                                                       |
| --------------------- | ----------------------------------------------------------------------------- |
| `np.random.seed(5)`   | Fix the random number generator to repeat the same sequence                   |
| `np.random.randn(10)` | Generate 10 random float numbers from **normal distribution** (mean=0, std=1) |

---


In [16]:
l1=[1,2,3,4,5,6,7,8,9] # A regular Python list
l=np.array(l1) # Convert it to a NumPy array
l # Display the array

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

Why use NumPy arrays?

Compared to Python lists, NumPy arrays:

- Are faster and more memory-efficient

- Support vectorized operations (e.g., l * 2 multiplies all elements by 2)

- Have built-in methods for mathematical, statistical, and matrix operations



In [17]:
min(l1) # returns the smallest element in the list.

1

If l is a NumPy array

In [18]:
l.min()


np.int64(1)

Both min(l1) and l.min() are valid — just depend on whether you're using a Python list or a NumPy array.

In [19]:
max(l1) # returns the largest element in the list.

9

If l is a NumPy array

In [20]:
l.max()


np.int64(9)

# reshape

In [5]:
import numpy as np
a=np.arange(81) # Creates a NumPy array with integers starting from 0 up to (but not including)
a

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80])

In [6]:
a.reshape(9,9) #It reshapes the 1D array a (which contains 81 elements) into a 2D array (matrix) with 9 rows and 9 columns.

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23, 24, 25, 26],
       [27, 28, 29, 30, 31, 32, 33, 34, 35],
       [36, 37, 38, 39, 40, 41, 42, 43, 44],
       [45, 46, 47, 48, 49, 50, 51, 52, 53],
       [54, 55, 56, 57, 58, 59, 60, 61, 62],
       [63, 64, 65, 66, 67, 68, 69, 70, 71],
       [72, 73, 74, 75, 76, 77, 78, 79, 80]])

In [3]:
b=np.arange(0,50,2)
b

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32,
       34, 36, 38, 40, 42, 44, 46, 48])

 `np.arange(start, stop, step)`

* **start = 0** → Starting value
* **stop = 50** → Ending **just before 50** (does not include 50)
* **step = 2** → Increment by 2 each time

---



In [4]:
type(b) #will return the type of the object 

numpy.ndarray


* `b` is assigned the result of `np.arange(0, 50, 2)`, which is a NumPy array.
* So, `type(b)` will return the type of the object `b`.

---

### 🧠 What does this mean?

* `b` is a **NumPy array object**, specifically of type `ndarray` (n-dimensional array).
* This is the basic type used in NumPy to store arrays of any shape and data type.

---

To check its shape, size, or data type too:

```python
b.shape     # shape of the array
b.size      # total number of elements
b.dtype     # data type of the elements
```


In [6]:
b[0] #This accesses the first element (index 0) of the NumPy array b.

np.int64(0)

In [7]:
b[3] #Accesses the element at index 3 (4th element, since indexing starts at 0)


np.int64(6)

In [8]:
b[-1] # -1 refers to the last element in the NumPy array b.

np.int64(48)

In [9]:
b[:2]

array([0, 2])


---

### 🔍 What it does:

* This is **slicing** the array `b`.
* It means: **"Get all elements from the start (index 0) up to (but not including) index 2."**

---

### ✅ Assuming:

```python
b = np.arange(0, 50, 2)
```

So the array `b` is:

```python
array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18,
       20, 22, 24, 26, 28, 30, 32, 34, 36, 38,
       40, 42, 44, 46, 48])
```

---

### 🎯 Output of `b[:2]`:

```python
array([0, 2])
```

* These are the elements at **index 0** and **index 1**

---

### 🧠 Summary

| Expression | Meaning                                   | Output          |
| ---------- | ----------------------------------------- | --------------- |
| `b[:2]`    | Slice from start to index 2 (excluding 2) | `array([0, 2])` |

also try:

* `b[2:5]` → elements at index 2, 3, and 4
* `b[::2]` → every second element
* `b[-3:]` → last 3 elements



In [10]:
b[:5]

array([0, 2, 4, 6, 8])

In [11]:
b[2:5]

array([4, 6, 8])

In [12]:
b[::2]

array([ 0,  4,  8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48])

In [13]:
b[-1]

np.int64(48)

In [14]:
b[-3]

np.int64(44)

In [15]:
b[-3:]

array([44, 46, 48])

In [16]:
b[:-3]

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32,
       34, 36, 38, 40, 42])

In [12]:
a=np.arange(0,20,2)
a

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

---

#### ✅ Function: `np.arange(start, stop, step)`

* **start = 0** → starting value
* **stop = 20** → stopping *before* 20
* **step = 2** → increment by 2

---

### 🧾 Output:

```python
array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])
```

* It includes **even numbers** from 0 up to **(but not including)** 20.
* Total elements = **10**

---

### 🧠 Summary

| Parameter | Value                     |
| --------- | ------------------------- |
| Start     | 0                         |
| Stop      | 20 (excluded)             |
| Step      | 2                         |
| Result    | Even numbers from 0 to 18 |

---


In [10]:
a[2:4]=5
a

array([0, 2, 5, 5, 5, 5, 5, 5, 5, 5])


---

### 🧾 Step 1: Initial array

```python
a = np.arange(0, 20, 2)
```

This creates:

```python
array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])
         ↑   ↑   ↑   ↑   ↑   ↑   ↑   ↑   ↑   ↑
Index:   0   1   2   3   4   5   6   7   8   9
```

---

### 🧾 Step 2: Modify slice

```python
a[2:4] = 5
```

* `a[2:4]` selects **index 2 and 3** → values `[4, 6]`
* `= 5` replaces both of them with **5**

---

### ✅ Final array:

```python
array([ 0,  2,  5,  5,  8, 10, 12, 14, 16, 18])
```

The values at index 2 and 3 have been **updated to 5**.

---

### 🧠 Summary

| Expression   | Meaning                                | Result            |
| ------------ | -------------------------------------- | ----------------- |
| `a[2:4] = 5` | Replace values at index 2 and 3 with 5 | `[4, 6] → [5, 5]` |

You can also assign different values using a list:

```python
a[2:4] = [50, 60]
```
---

In [6]:
a[0:2]

array([0, 2])


---

### 🔍 What it does:

This is **slicing** the NumPy array `a`.

It means:
➡️ **Select elements starting from index `0` up to (but not including) index `2`**.

---

### 🧾 Assuming the current array is:

From your previous step:

```python
a = array([ 0,  2,  5,  5,  8, 10, 12, 14, 16, 18])
```

---

### 🎯 Output of `a[0:2]`:

```python
array([0, 2])
```

Because:

* `a[0]` → 0
* `a[1]` → 2
* `a[2]` is **not** included (upper bound is exclusive)

---

### 🧠 Summary

| Expression | Meaning                 | Output          |
| ---------- | ----------------------- | --------------- |
| `a[0:2]`   | Slice from index 0 to 1 | `array([0, 2])` |


In [13]:
sliced=a[0:2].copy()
sliced

array([0, 2])


---

### 🔍 Step-by-Step Explanation

#### ✅ `a[0:2]`

* Slices the array `a` to get the elements at **index 0 and 1**
* Example: if `a = [0, 2, 5, 5, 8, 10, 12, 14, 16, 18]`, then
  `a[0:2]` returns `[0, 2]`

---

#### ✅ `.copy()`

* This creates a **new independent copy** of the sliced array.
* Without `.copy()`, any changes to `sliced` could affect the original array `a`.
* With `.copy()`, `sliced` becomes a **separate array in memory**.

---


### 🧠 Why use `.copy()`?

| Without `.copy()`             | With `.copy()`                            |
| ----------------------------- | ----------------------------------------- |
| `sliced` is a **view** of `a` | `sliced` is a **copy** of `a`             |
| Changing `sliced` affects `a` | Changing `sliced` does **not** affect `a` |
| Memory-efficient              | Safer for independent changes             |

---



In [14]:
a

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [15]:
sliced

array([0, 2])

In [17]:
l=np.arange(1,100,4)
l

array([ 1,  5,  9, 13, 17, 21, 25, 29, 33, 37, 41, 45, 49, 53, 57, 61, 65,
       69, 73, 77, 81, 85, 89, 93, 97])


---

### 🔍 Explanation of `np.arange(1, 100, 4)`

#### ✅ Syntax:

```python
np.arange(start, stop, step)
```

* **start = 1** → Start from 1
* **stop = 100** → Go up to **but not including** 100
* **step = 4** → Increment by 4

---

### ✅ What this generates:

A NumPy array starting at 1, adding 4 repeatedly, and stopping before 100.

```python
array([ 1,  5,  9, 13, 17, 21, 25, 29, 33, 37,
       41, 45, 49, 53, 57, 61, 65, 69, 73, 77,
       81, 85, 89, 93, 97])
```

* Total elements: **25**

---

### 🧠 Summary

| Parameter | Value                                            |
| --------- | ------------------------------------------------ |
| Start     | 1                                                |
| Stop      | 100 (excluded)                                   |
| Step      | 4                                                |
| Result    | Array of numbers increasing by 4 from 1 up to 97 |


In [19]:
sliced=l[0:7].copy()
sliced

array([ 1,  5,  9, 13, 17, 21, 25])



#### ✅ `l[0:7]`

* Slices the array `l` from index **0 up to (but not including) 7**
* This means it extracts the **first 7 elements**

#### ✅ `.copy()`

* Makes an **independent copy** of that slice
* Changes to `sliced` will **not affect** the original array `l`

---

### 🔢 Assuming `l` is:

```python
l = np.arange(1, 100, 4)
```

Then:

```python
l = array([ 1,  5,  9, 13, 17, 21, 25, 29, 33, 37,
           41, 45, 49, 53, 57, 61, 65, 69, 73, 77,
           81, 85, 89, 93, 97])
```

---

### ✅ Output of `sliced`:

```python
array([ 1,  5,  9, 13, 17, 21, 25])
```

---

### 🧠 Summary

| Expression | Meaning                              |
| ---------- | ------------------------------------ |
| `l[0:7]`   | First 7 elements from array `l`      |
| `.copy()`  | Creates a separate copy of the slice |
| `sliced`   | `[1, 5, 9, 13, 17, 21, 25]`          |

---


In [20]:
sliced[::]=10
sliced

array([10, 10, 10, 10, 10, 10, 10])


---

### 🔍 Explanation

#### ✅ `sliced[::]`

* This is **slicing the entire array**.
* It's equivalent to `sliced[:]` or just `sliced` — it selects **all elements**.
* The general format is `sliced[start:stop:step]`.

  * `::` means: from beginning to end, step by 1.

#### ✅ `= 10`

* This assigns the value **10 to all selected elements**.

---

### 💡 Result:

Before:

```python
sliced = [ 1,  5,  9, 13, 17, 21, 25]
```

After:

```python
sliced = [10, 10, 10, 10, 10, 10, 10]
```

---

### 🧠 Summary

| Expression        | Meaning                      | Result                         |
| ----------------- | ---------------------------- | ------------------------------ |
| `sliced[::] = 10` | Replace all elements with 10 | `[10, 10, 10, 10, 10, 10, 10]` |

✅ Since you used `.copy()` earlier, the **original array `l` remains unchanged**.

---

# 2D array

In [24]:
# Create a 2D NumPy array (3x3 matrix)
list_2d = np.array([
    [1, 2, 3],   # First row
    [3, 4, 5],   # Second row
    [5, 6, 7]    # Third row
])

list_2d

array([[1, 2, 3],
       [3, 4, 5],
       [5, 6, 7]])

In [25]:
list_2d.shape    # Returns (3, 3) → shape of the array (3 rows, 3 columns)

(3, 3)

In [32]:
list_2d.ndim     # Returns 2 → the array is 2-dimensional

2

In [33]:
list_2d.size     # Returns 9 → total number of elements in the array

9

In [34]:
list_2d[0][1]    # Returns 2 → element at row 0, column 1

np.int64(2)

In [35]:
list_2d[1][2] # Returns 5 → another way to access element at row 1, column 2


np.int64(5)

In [36]:
list_2d[1][1]

np.int64(4)

In [37]:
a=np.array([[1,2,4],[3,5,6],[4,5,9],[5,5,3]])
a

array([[1, 2, 4],
       [3, 5, 6],
       [4, 5, 9],
       [5, 5, 3]])

In [38]:
a[1:] # Slicing the array to get all rows starting from index 1 to the end

array([[3, 5, 6],
       [4, 5, 9],
       [5, 5, 3]])

In [39]:
# Slice the array:
# a[1:, 1:] means:
# - From row index 1 to the end → rows 1, 2, and 3
# - From column index 1 to the end → columns 1 and 2

a[1:,1:]

array([[5, 6],
       [5, 9],
       [5, 3]])

In [40]:
a[a>2] # This line selects all elements in 'a' that are greater than 2

array([4, 3, 5, 6, 4, 5, 9, 5, 5, 3])

In [41]:
a[a<3]
# Select all elements from 'a' that are less than 3

array([1, 2])

In [42]:
x=np.arange(1,30,4) # It starts at 1 and adds 4 repeatedly until just before 30
x

array([ 1,  5,  9, 13, 17, 21, 25, 29])

In [43]:
x+2 # Add 2 to every element in the array

array([ 3,  7, 11, 15, 19, 23, 27, 31])

In [44]:
x-2 # Subtract 2 from each element in the array

array([-1,  3,  7, 11, 15, 19, 23, 27])

In [45]:
x*3 # Multiply each element in the array by 3

array([ 3, 15, 27, 39, 51, 63, 75, 87])

In [46]:
x/3 # Divide each element by 3

array([0.33333333, 1.66666667, 3.        , 4.33333333, 5.66666667,
       7.        , 8.33333333, 9.66666667])

In [47]:
y=np.arange(3,30) # Create an array starting from 3 to 29 (30 is excluded)
y

array([ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
       20, 21, 22, 23, 24, 25, 26, 27, 28, 29])

In [48]:
np.sqrt(y) # Calculates square root of each element in y

array([1.73205081, 2.        , 2.23606798, 2.44948974, 2.64575131,
       2.82842712, 3.        , 3.16227766, 3.31662479, 3.46410162,
       3.60555128, 3.74165739, 3.87298335, 4.        , 4.12310563,
       4.24264069, 4.35889894, 4.47213595, 4.58257569, 4.69041576,
       4.79583152, 4.89897949, 5.        , 5.09901951, 5.19615242,
       5.29150262, 5.38516481])

In [49]:
np.cos(y) # Cosine of each value in y

array([-0.9899925 , -0.65364362,  0.28366219,  0.96017029,  0.75390225,
       -0.14550003, -0.91113026, -0.83907153,  0.0044257 ,  0.84385396,
        0.90744678,  0.13673722, -0.75968791, -0.95765948, -0.27516334,
        0.66031671,  0.98870462,  0.40808206, -0.54772926, -0.99996083,
       -0.53283302,  0.42417901,  0.99120281,  0.64691932, -0.29213881,
       -0.96260587, -0.74805753])

In [50]:
np.sin(y) # Calculate sine of each value (in radians)

array([ 0.14112001, -0.7568025 , -0.95892427, -0.2794155 ,  0.6569866 ,
        0.98935825,  0.41211849, -0.54402111, -0.99999021, -0.53657292,
        0.42016704,  0.99060736,  0.65028784, -0.28790332, -0.96139749,
       -0.75098725,  0.14987721,  0.91294525,  0.83665564, -0.00885131,
       -0.8462204 , -0.90557836, -0.13235175,  0.76255845,  0.95637593,
        0.27090579, -0.66363388])

In [51]:
np.tan(y) # Compute tangent of each element (in radians)

array([-1.42546543e-01,  1.15782128e+00, -3.38051501e+00, -2.91006191e-01,
        8.71447983e-01, -6.79971146e+00, -4.52315659e-01,  6.48360827e-01,
       -2.25950846e+02, -6.35859929e-01,  4.63021133e-01,  7.24460662e+00,
       -8.55993401e-01,  3.00632242e-01,  3.49391565e+00, -1.13731371e+00,
        1.51589471e-01,  2.23716094e+00, -1.52749853e+00,  8.85165604e-03,
        1.58815308e+00, -2.13489670e+00, -1.33526407e-01,  1.17875355e+00,
       -3.27370380e+00, -2.81429605e-01,  8.87142844e-01])

In [52]:
y.max() # Finds the maximum value in the array

np.int64(29)

In [53]:
y.min() # Find the minimum value in the array

np.int64(3)

In [ ]:
y.var()

# Calculate the variance of the array
# Variance measures how spread out the values are
# Formula used: mean((x - mean(x))²)

60.666666666666664

In [54]:
y.std()

# Compute the standard deviation of the array 
# Standard deviation = sqrt(variance)


np.float64(7.788880963698615)

In [55]:
list_2D=([[2,3,4],[4,3,2],[4,5,6]])
list_2D

[[2, 3, 4], [4, 3, 2], [4, 5, 6]]

In [56]:
L=np.array(list_2D) # Converting the 2D list to a NumPy array
L

array([[2, 3, 4],
       [4, 3, 2],
       [4, 5, 6]])

In [ ]:
L.sum(axis=0)

# sums the elements column-wise (i.e., down each column).

# axis=0 means "operate across rows", so it will add values in each column.

array([10, 11, 12])

| Column Index | Values    | Sum |
| ------------ | --------- | --- |
| 0            | 2 + 4 + 4 | 10  |
| 1            | 3 + 3 + 5 | 11  |
| 2            | 4 + 2 + 6 | 12  |


In [ ]:
L.sum(axis=1)

# axis=1 means row-wise operation.

# So this adds the elements across each row (i.e., left to right in each row).

array([ 9,  9, 15])

| Row Index | Values    | Sum |
| --------- | --------- | --- |
| 0         | 2 + 3 + 4 | 9   |
| 1         | 4 + 3 + 2 | 9   |
| 2         | 4 + 5 + 6 | 15  |
